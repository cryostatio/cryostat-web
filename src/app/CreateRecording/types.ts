/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
