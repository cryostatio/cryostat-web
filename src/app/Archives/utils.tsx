/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
