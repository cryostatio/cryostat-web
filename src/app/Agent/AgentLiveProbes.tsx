/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ErrorView } from '@app/ErrorView/ErrorView';
import { authFailMessage, isAuthFail } from '@app/ErrorView/types';
import { DeleteWarningModal } from '@app/Modal/DeleteWarningModal';
import { DeleteOrDisableWarningType } from '@app/Modal/types';
import { LoadingView } from '@app/Shared/Components/LoadingView';
import { LoadingProps } from '@app/Shared/Components/types';
import { EventProbe, NotificationCategory } from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { sortResources, TableColumn } from '@app/utils/utils';
import {
  Button,
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem,
  TextInput,
  Stack,
  StackItem,
  EmptyState,
  EmptyStateIcon,
  EmptyStateHeader,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import {
  TableVariant,
  ISortBy,
  SortByDirection,
  ThProps,
  Table,
  Tbody,
  Th,
  Thead,
  Tr,
  Td,
} from '@patternfly/react-table';
import * as React from 'react';
import { combineLatest } from 'rxjs';
import { AboutAgentCard } from './AboutAgentCard';

export type LiveProbeActions = 'REMOVE';

const tableColumns: TableColumn[] = [
  {
    title: 'ID',
    keyPaths: ['id'],
    sortable: true,
  },
  {
    title: 'Name',
    keyPaths: ['name'],
    sortable: true,
  },
  {
    title: 'Class',
    keyPaths: ['clazz'],
    sortable: true,
  },
  {
    title: 'Description',
    keyPaths: ['description'],
    sortable: true,
  },
  {
    title: 'Method',
    keyPaths: ['methodName'],
    sortable: true,
  },
];

export interface AgentLiveProbesProps {}

export const AgentLiveProbes: React.FC<AgentLiveProbesProps> = (_) => {
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();

  const [probes, setProbes] = React.useState<EventProbe[]>([]);
  const [filteredProbes, setFilteredProbes] = React.useState<EventProbe[]>([]);
  const [filterText, setFilterText] = React.useState('');
  const [sortBy, setSortBy] = React.useState<ISortBy>({});
  const [isLoading, setIsLoading] = React.useState(false);
  const [errorMessage, setErrorMessage] = React.useState('');
  const [warningModalOpen, setWarningModalOpen] = React.useState(false);
  const [actionLoadings, setActionLoadings] = React.useState<Record<LiveProbeActions, boolean>>({ REMOVE: false });

  const getSortParams = React.useCallback(
    (columnIndex: number): ThProps['sort'] => ({
      sortBy: sortBy,
      onSort: (_event, index, direction) => {
        setSortBy({
          index: index,
          direction: direction,
        });
      },
      columnIndex,
    }),
    [sortBy, setSortBy],
  );

  const handleProbes = React.useCallback(
    (probes) => {
      setProbes(probes);
      setErrorMessage('');
      setIsLoading(false);
    },
    [setProbes, setIsLoading, setErrorMessage],
  );

  const handleError = React.useCallback(
    (error) => {
      setErrorMessage(error.message);
      setIsLoading(false);
    },
    [setIsLoading, setErrorMessage],
  );

  const refreshProbes = React.useCallback(() => {
    setIsLoading(true);
    addSubscription(
      context.api.getActiveProbes(true).subscribe({
        next: (value) => handleProbes(value),
        error: (err) => handleError(err),
      }),
    );
  }, [addSubscription, context.api, setIsLoading, handleProbes, handleError]);

  const authRetry = React.useCallback(() => {
    context.target.setAuthRetry();
  }, [context.target]);

  const handleDeleteAllProbes = React.useCallback(() => {
    setActionLoadings((old) => {
      return {
        ...old,
        REMOVE: true,
      };
    });
    addSubscription(
      context.api.removeProbes().subscribe(() => {
        setActionLoadings((old) => {
          return {
            ...old,
            REMOVE: false,
          };
        });
      }),
    );
  }, [addSubscription, context.api, setActionLoadings]);

  const handleWarningModalAccept = React.useCallback(() => handleDeleteAllProbes(), [handleDeleteAllProbes]);

  const handleWarningModalClose = React.useCallback(() => {
    setWarningModalOpen(false);
  }, [setWarningModalOpen]);

  const handleDeleteButton = React.useCallback(() => {
    if (context.settings.deletionDialogsEnabledFor(DeleteOrDisableWarningType.DeleteActiveProbes)) {
      setWarningModalOpen(true);
    } else {
      handleDeleteAllProbes();
    }
  }, [context.settings, setWarningModalOpen, handleDeleteAllProbes]);

  const handleFilterTextChange = React.useCallback((_, value: string) => setFilterText(value), [setFilterText]);

  React.useEffect(() => {
    addSubscription(
      context.target.target().subscribe(() => {
        setFilterText('');
        refreshProbes();
      }),
    );
  }, [context, context.target, addSubscription, setFilterText, refreshProbes]);

  React.useEffect(() => {
    if (!context.settings.autoRefreshEnabled()) {
      return;
    }
    const id = window.setInterval(
      () => refreshProbes(),
      context.settings.autoRefreshPeriod() * context.settings.autoRefreshUnits(),
    );
    return () => window.clearInterval(id);
  }, [context.settings, refreshProbes]);

  React.useEffect(() => {
    addSubscription(
      context.target.authFailure().subscribe(() => {
        setErrorMessage(authFailMessage);
      }),
    );
  }, [addSubscription, context.target, setErrorMessage]);

  React.useEffect(() => {
    addSubscription(
      combineLatest([
        context.target.target(),
        context.notificationChannel.messages(NotificationCategory.ProbeTemplateApplied),
      ]).subscribe(([currentTarget, e]) => {
        if (currentTarget?.jvmId != e.message.jvmId) {
          return;
        }
        setProbes((old) => {
          const probes = e.message.events as EventProbe[];
          const probeIds = probes.map((p) => p.id);
          if (probes?.length > 0) {
            return [...old.filter((probe) => !probeIds.includes(probe.id)), ...probes];
          }
          return old;
        });
      }),
    );
  }, [addSubscription, context, context.notificationChannel, context.target, setProbes]);

  React.useEffect(() => {
    addSubscription(
      combineLatest([
        context.target.target(),
        context.notificationChannel.messages(NotificationCategory.ProbesRemoved),
      ]).subscribe(([currentTarget, e]) => {
        if (currentTarget?.jvmId == e.message.jvmId) {
          setProbes([]);
        }
      }),
    );
  }, [addSubscription, context, context.notificationChannel, setProbes]);

  React.useEffect(() => {
    let filtered: EventProbe[];
    if (!filterText) {
      filtered = probes;
    } else {
      const ft = filterText.trim().toLowerCase();
      filtered = probes.filter(
        (t: EventProbe) =>
          t.name.toLowerCase().includes(ft) ||
          t.description.toLowerCase().includes(ft) ||
          t.clazz.toLowerCase().includes(ft) ||
          t.methodDescriptor.toLowerCase().includes(ft) ||
          t.methodName.toLowerCase().includes(ft),
      );
    }

    setFilteredProbes(
      sortResources(
        {
          index: sortBy.index ?? 0,
          direction: sortBy.direction ?? SortByDirection.asc,
        },
        filtered,
        tableColumns,
      ),
    );
  }, [filterText, probes, sortBy, setFilteredProbes]);

  const probeRows = React.useMemo(
    () =>
      filteredProbes.map((t: EventProbe, index) => (
        <Tr key={`active-probe-${index}`}>
          <Td key={`active-probe-id-${index}`} dataLabel={tableColumns[0].title}>
            {t.id}
          </Td>
          <Td key={`active-probe-name-${index}`} dataLabel={tableColumns[1].title}>
            {t.name}
          </Td>
          <Td key={`active-probe-clazz-${index}`} dataLabel={tableColumns[2].title}>
            {t.clazz}
          </Td>
          <Td key={`active-probe-description-${index}`} dataLabel={tableColumns[3].title}>
            {t.description}
          </Td>
          <Td key={`active-probe-methodname-${index}`} dataLabel={tableColumns[4].title}>
            {t.methodName}
          </Td>
        </Tr>
      )),
    [filteredProbes],
  );

  const actionLoadingProps = React.useMemo<Record<LiveProbeActions, LoadingProps>>(
    () => ({
      REMOVE: {
        spinnerAriaValueText: 'Removing',
        spinnerAriaLabel: 'removing-all-probes',
        isLoading: actionLoadings['REMOVE'],
      } as LoadingProps,
    }),
    [actionLoadings],
  );

  if (errorMessage != '') {
    return (
      <ErrorView
        title={'Error retrieving active probes'}
        message={`${errorMessage}. Note: This error is generally caused when the agent is not loaded or is not active. Check that the target was started with the agent (-javaagent:/path/to/agent.jar).`}
        retry={isAuthFail(errorMessage) ? authRetry : undefined}
      />
    );
  } else if (isLoading) {
    return <LoadingView />;
  } else {
    return (
      <>
        <Stack hasGutter style={{ marginTop: '1em' }}>
          <StackItem>
            <AboutAgentCard />
          </StackItem>
          <StackItem>
            <Toolbar id="active-probes-toolbar">
              <ToolbarContent>
                <ToolbarGroup variant="filter-group">
                  <ToolbarItem>
                    <TextInput
                      name="activeProbeFilter"
                      id="activeProbeFilter"
                      type="search"
                      placeholder="Filter..."
                      aria-label="Active probe filter"
                      onChange={handleFilterTextChange}
                    />
                  </ToolbarItem>
                </ToolbarGroup>
                <ToolbarGroup variant="icon-button-group">
                  <ToolbarItem>
                    <Button
                      key="delete"
                      variant="danger"
                      onClick={handleDeleteButton}
                      isDisabled={!filteredProbes.length || actionLoadings['REMOVE']}
                      {...actionLoadingProps['REMOVE']}
                    >
                      {actionLoadings['REMOVE'] ? 'Removing' : 'Remove'} all probes
                    </Button>
                  </ToolbarItem>
                </ToolbarGroup>
              </ToolbarContent>
              <DeleteWarningModal
                warningType={DeleteOrDisableWarningType.DeleteActiveProbes}
                visible={warningModalOpen}
                onAccept={handleWarningModalAccept}
                onClose={handleWarningModalClose}
              />
            </Toolbar>
            {probeRows.length ? (
              <Table aria-label="Active probe table" variant={TableVariant.compact}>
                <Thead>
                  <Tr>
                    {tableColumns.map(({ title, sortable }, index) => (
                      <Th key={`active-probe-header-${title}`} sort={sortable ? getSortParams(index) : undefined}>
                        {title}
                      </Th>
                    ))}
                  </Tr>
                </Thead>
                <Tbody>{probeRows}</Tbody>
              </Table>
            ) : (
              <EmptyState>
                <EmptyStateHeader
                  titleText="No active probes"
                  icon={<EmptyStateIcon icon={SearchIcon} />}
                  headingLevel="h4"
                />
              </EmptyState>
            )}
          </StackItem>
        </Stack>
      </>
    );
  }
};
