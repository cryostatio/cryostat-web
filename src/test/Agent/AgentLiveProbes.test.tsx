/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AgentLiveProbes } from '@app/Agent/AgentLiveProbes';
import { DeleteActiveProbes } from '@app/Modal/types';
import {
  MessageType,
  EventProbe,
  NotificationCategory,
  MessageMeta,
  NotificationMessage,
} from '@app/Shared/Services/api.types';
import { defaultServices } from '@app/Shared/Services/Services';
import '@testing-library/jest-dom';
import { act, cleanup, screen, within } from '@testing-library/react';
import { of } from 'rxjs';
import { render, renderSnapshot, testT } from '../utils';

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockJvmId = 'id';
const mockTarget = {
  agent: false,
  connectUrl: mockConnectUrl,
  alias: 'fooTarget',
  jvmId: mockJvmId,
  labels: [],
  annotations: { cryostat: [], platform: [] },
};

const mockMessageType = { type: 'application', subtype: 'json' } as MessageType;

const mockProbe: EventProbe = {
  id: 'some_id',
  name: 'some_name',
  clazz: 'some_clazz',
  description: 'some_desc',
  recordStackTrace: true,
  useRethrow: true,
  methodName: 'a_method',
  methodDescriptor: 'method_desc',
  location: 'some_loc',
  returnValue: 'a_value',
  parameters: 'some_params',
  fields: 'some_fields',
  path: 'some_path',
};

const mockAnotherProbe: EventProbe = {
  ...mockProbe,
  id: 'another_id',
  name: 'another_name',
};

const mockApplyTemplateNotification = {
  meta: {
    category: NotificationCategory.ProbeTemplateApplied,
    type: mockMessageType,
  } as MessageMeta,
  message: {
    jvmId: mockJvmId,
    events: [mockAnotherProbe],
  },
} as NotificationMessage;

const mockRemoveProbesNotification = {
  meta: {
    category: NotificationCategory.ProbesRemoved,
    type: mockMessageType,
  } as MessageMeta,
  message: {
    jvmId: mockJvmId,
  },
} as NotificationMessage;

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest.spyOn(defaultServices.target, 'authFailure').mockReturnValue(of());

jest
  .spyOn(defaultServices.settings, 'deletionDialogsEnabledFor')
  .mockReturnValueOnce(false) // should remove all probes when Remove All Probe is clicked
  .mockReturnValue(true); // should show warning modal and remove all probes when confirmed

jest
  .spyOn(defaultServices.api, 'getActiveProbes')
  // .mockReturnValueOnce(of([mockProbe])) // renders correctly

  .mockReturnValueOnce(of([])) // should disable remove button if there is no probe

  .mockReturnValueOnce(of([mockProbe])) // should add a probe after receiving a notification

  .mockReturnValue(of([mockProbe])); // All other tests

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  // .mockReturnValueOnce(of()) // renders correctly
  // .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // should disable remove button if there is no probe
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of(mockApplyTemplateNotification)) // should add a probe after receiving a notification
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockRemoveProbesNotification)) // should remove a probe after receiving a notification

  .mockReturnValue(of()); // All other tests

describe('<AgentLiveProbes />', () => {
  afterEach(cleanup);

  it.skip('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('should disable remove button if there is no probe', async () => {
    render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    const removeButton = screen.getByText('Remove all probes');
    expect(removeButton).toBeInTheDocument();
    expect(removeButton).toBeVisible();
    expect(removeButton).toBeDisabled();
  });

  it('should add a probe after receiving a notification', async () => {
    render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    const addTemplateName = screen.getByText('another_name');
    expect(addTemplateName).toBeInTheDocument();
    expect(addTemplateName).toBeVisible();
  });

  it('should remove all probes after receiving a notification', async () => {
    render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    expect(screen.queryByText('some_name')).not.toBeInTheDocument();
    expect(screen.queryByText('another_name')).not.toBeInTheDocument();
  });

  it('should display the column header fields', async () => {
    render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    const headers = ['ID', 'Name', 'Class', 'Description', 'Method'];
    headers.forEach((header) => {
      const nameHeader = screen.getByText(header);
      expect(nameHeader).toBeInTheDocument();
      expect(nameHeader).toBeVisible();
    });
  });

  it('should remove all probes when remove all probe is clicked', async () => {
    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'removeProbes').mockReturnValue(of(true));
    const { user } = render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    const removeButton = screen.getByText('Remove all probes');
    expect(removeButton).toBeInTheDocument();
    expect(removeButton).toBeVisible();

    await user.click(removeButton);

    expect(deleteRequestSpy).toBeCalledTimes(1);
  });

  it('should show warning modal and remove all probes when confirmed', async () => {
    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'removeProbes').mockReturnValue(of(true));
    const { user } = render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    await act(async () => {
      const removeButton = screen.getByText('Remove all probes');
      expect(removeButton).toBeInTheDocument();
      expect(removeButton).toBeVisible();

      await user.click(removeButton);

      const warningModal = await screen.findByRole('dialog');
      expect(warningModal).toBeInTheDocument();
      expect(warningModal).toBeVisible();

      const modalTitle = within(warningModal).getByText(DeleteActiveProbes.title);
      expect(modalTitle).toBeInTheDocument();
      expect(modalTitle).toBeVisible();

      const confirmButton = within(warningModal).getByText('Delete');
      expect(confirmButton).toBeInTheDocument();
      expect(confirmButton).toBeVisible();

      await user.click(confirmButton);
    });

    expect(deleteRequestSpy).toBeCalledTimes(1);
  });

  it('should shown empty state when table is empty', async () => {
    const { user } = render({ routerConfigs: { routes: [{ path: '/events', element: <AgentLiveProbes /> }] } });

    const filterInput = screen.getByLabelText(testT('AgentLiveProves.ARIA_LABELS.SEARCH_INPUT'));
    expect(filterInput).toBeInTheDocument();
    expect(filterInput).toBeVisible();

    await user.type(filterInput, 'someveryoddname');

    expect(screen.queryByText('some_name')).not.toBeInTheDocument();

    const hintText = screen.getByText('No active probes');
    expect(hintText).toBeInTheDocument();
    expect(hintText).toBeVisible();
  });
});
