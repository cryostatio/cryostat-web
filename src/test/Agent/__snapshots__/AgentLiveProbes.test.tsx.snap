// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AgentLiveProbes /> renders correctly 1`] = `
<div
  className="pf-l-stack pf-m-gutter"
  style={
    Object {
      "marginTop": "1em",
    }
  }
>
  <div
    className="pf-l-stack__item"
  >
    <article
      className="pf-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF4/Card"
      data-ouia-safe={true}
      id=""
    >
      <div
        className="pf-c-card__title"
      >
        About the JMC Agent
      </div>
      <div
        className="pf-c-card__body"
      >
        The JMC Agent allows users to dynamically inject custom JFR events into running JVMs. In order to make use of the JMC Agent, the agent jar must be present in the same container as the target, and the target must be started with the agent (-javaagent:/path/to/agent.jar). Once these pre-requisites are met, the user can upload probe templates to Cryostat and insert them to the target, as well as view or remove currently active probes.
      </div>
    </article>
  </div>
  <div
    className="pf-l-stack__item"
  >
    <div
      className="pf-c-toolbar"
      data-ouia-component-id="OUIA-Generated-Toolbar-1"
      data-ouia-component-type="PF4/Toolbar"
      data-ouia-safe={true}
      id="active-probes-toolbar"
    >
      <div
        className="pf-c-toolbar__content"
      >
        <div
          className="pf-c-toolbar__content-section"
        >
          <div
            className="pf-c-toolbar__group pf-m-filter-group"
          >
            <div
              className="pf-c-toolbar__item"
            >
              <input
                aria-invalid={false}
                aria-label="Active probe filter"
                className="pf-c-form-control"
                data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                data-ouia-component-type="PF4/TextInput"
                data-ouia-safe={true}
                disabled={false}
                id="activeProbeFilter"
                name="activeProbeFilter"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Filter..."
                required={false}
                type="search"
              />
            </div>
          </div>
          <div
            className="pf-c-toolbar__group pf-m-icon-button-group"
          >
            <div
              className="pf-c-toolbar__item"
            >
              <button
                aria-disabled={false}
                aria-label={null}
                className="pf-c-button pf-m-danger pf-m-progress"
                data-ouia-component-id="OUIA-Generated-Button-danger-1"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                Remove
                 All Probes
              </button>
            </div>
          </div>
        </div>
        <div
          className="pf-c-toolbar__expandable-content"
          id="active-probes-toolbar-expandable-content-2"
        >
          <div
            className="pf-c-toolbar__group"
          />
        </div>
      </div>
      <div
        className="pf-c-toolbar__content pf-m-hidden"
        hidden={true}
      >
        <div
          className="pf-c-toolbar__group"
        />
      </div>
    </div>
    <table
      aria-label="Active Probe Table"
      className="pf-c-table pf-m-grid-md pf-m-compact"
      data-ouia-component-id="OUIA-Generated-Table-1"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe={true}
      role="grid"
    >
      <thead
        className=""
      >
        <tr
          className=""
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe={true}
          hidden={false}
        >
          <th
            aria-sort="none"
            className="pf-c-table__sort"
            onMouseEnter={[Function]}
            scope="col"
          >
            <button
              className="pf-c-table__button"
              onClick={[Function]}
              type="button"
            >
              <div
                className="pf-c-table__button-content"
              >
                <span
                  className="pf-c-table__text"
                  onMouseEnter={[Function]}
                >
                  ID
                </span>
                <span
                  className="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
          <th
            aria-sort="none"
            className="pf-c-table__sort"
            onMouseEnter={[Function]}
            scope="col"
          >
            <button
              className="pf-c-table__button"
              onClick={[Function]}
              type="button"
            >
              <div
                className="pf-c-table__button-content"
              >
                <span
                  className="pf-c-table__text"
                  onMouseEnter={[Function]}
                >
                  Name
                </span>
                <span
                  className="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
          <th
            aria-sort="none"
            className="pf-c-table__sort"
            onMouseEnter={[Function]}
            scope="col"
          >
            <button
              className="pf-c-table__button"
              onClick={[Function]}
              type="button"
            >
              <div
                className="pf-c-table__button-content"
              >
                <span
                  className="pf-c-table__text"
                  onMouseEnter={[Function]}
                >
                  Class
                </span>
                <span
                  className="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
          <th
            aria-sort="none"
            className="pf-c-table__sort"
            onMouseEnter={[Function]}
            scope="col"
          >
            <button
              className="pf-c-table__button"
              onClick={[Function]}
              type="button"
            >
              <div
                className="pf-c-table__button-content"
              >
                <span
                  className="pf-c-table__text"
                  onMouseEnter={[Function]}
                >
                  Description
                </span>
                <span
                  className="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
          <th
            aria-sort="none"
            className="pf-c-table__sort"
            onMouseEnter={[Function]}
            scope="col"
          >
            <button
              className="pf-c-table__button"
              onClick={[Function]}
              type="button"
            >
              <div
                className="pf-c-table__button-content"
              >
                <span
                  className="pf-c-table__text"
                  onMouseEnter={[Function]}
                >
                  Method
                </span>
                <span
                  className="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        className=""
        role="rowgroup"
      >
        <tr
          className=""
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe={true}
          hidden={false}
        >
          <td
            className=""
            data-label="ID"
            onMouseEnter={[Function]}
          >
            some_id
          </td>
          <td
            className=""
            data-label="Name"
            onMouseEnter={[Function]}
          >
            some_name
          </td>
          <td
            className=""
            data-label="Class"
            onMouseEnter={[Function]}
          >
            some_clazz
          </td>
          <td
            className=""
            data-label="Description"
            onMouseEnter={[Function]}
          >
            some_desc
          </td>
          <td
            className=""
            data-label="Method"
            onMouseEnter={[Function]}
          >
            a_method
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
